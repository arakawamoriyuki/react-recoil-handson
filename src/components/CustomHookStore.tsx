import React from 'react';

export const useCounter = (initialCount = 0) => {
  const [count, setCount] = React.useState(initialCount);

  const increment = () => {
    setCount((count) => count + 1);
  };

  const reset = () => {
    setCount(0);
  };

  return { count, increment, reset };
}
