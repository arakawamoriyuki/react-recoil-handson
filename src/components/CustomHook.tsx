import React from 'react';
import Box from '@mui/material/Box';
import Typography from '@mui/material/Typography';
import Button from '@mui/material/Button';
import { useCounter } from './CustomHookStore';

const CustomHookComponent: React.FC = () => {
  const { count, increment, reset } = useCounter(0);

  return (
    <Box>
      <Typography>Count: {count}</Typography>
      <Button onClick={() => increment()}>+</Button>
      <Button onClick={() => reset()}>Reset</Button>
    </Box>
  );
};

export default CustomHookComponent;
