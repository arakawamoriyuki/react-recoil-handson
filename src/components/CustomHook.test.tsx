import { cleanup, render, screen, fireEvent } from '@testing-library/react';
import CustomHook from './CustomHook';
import * as CustomHookStoreModule from './CustomHookStore';

afterEach(() => cleanup());

describe('CustomHook', () => {
  it('CustomHookが表示される', () => {
    const { asFragment } = render(
      <CustomHook />,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('+ボタンが押されるとincrementが呼ばれる', () => {
    const mockIncrement = jest.fn();
    jest.spyOn(CustomHookStoreModule, 'useCounter').mockReturnValueOnce({
      count: 1,
      increment: mockIncrement,
      reset: () => {},
    });

    render(
      <CustomHook />,
    );

    const button = screen.getByRole('button', { name: '+' });
    fireEvent.click(button);

    expect(mockIncrement).toBeCalled();
  });

  it('Resetボタンが押されるとresetが呼ばれる', () => {
    const mockReset = jest.fn();
    jest.spyOn(CustomHookStoreModule, 'useCounter').mockReturnValueOnce({
      count: 1,
      increment: () => {},
      reset: mockReset,
    });

    render(
      <CustomHook />,
    );

    const button = screen.getByRole('button', { name: 'Reset' });
    fireEvent.click(button);

    expect(mockReset).toBeCalled();
  });
});
