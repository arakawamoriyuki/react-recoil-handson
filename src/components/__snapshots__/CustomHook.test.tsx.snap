// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomHook CustomHookが表示される 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-0"
  >
    <p
      class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
    >
      Count: 0
    </p>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1e6y48t-MuiButtonBase-root-MuiButton-root"
      tabindex="0"
      type="button"
    >
      +
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1e6y48t-MuiButtonBase-root-MuiButton-root"
      tabindex="0"
      type="button"
    >
      Reset
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </div>
</DocumentFragment>
`;
